app/metal-controller-manager/api/v1alpha1/serverclass_filter.go

```diff
@@ -4,7 +4,38 @@
 
 package v1alpha1
 
-import "sort"
+import (
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// FilterServers returns the subset of servers that:
+// - Are accepted.
+// - Match the ServerClass selection criteria.
+//
+// In case of error the returned slice will be nil.
+func (sc *ServerClass) FilterServers(servers []Server) ([]Server, error) {
+	s, err := metav1.LabelSelectorAsSelector(sc.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+
+	matches := make([]Server, 0, len(servers))
+
+	for _, server := range servers {
+		if !server.Spec.Accepted {
+			continue
+		}
+
+		if sc.Spec.Selector == nil || s.Matches(labels.Set(server.GetLabels())) {
+			matches = append(matches, server)
+		}
+	}
+
+	return FilterAcceptedServers(matches, sc.Spec.Qualifiers), nil
```

Comment:
We should un-export this method if it is no longer used directly.

Reply:
and we should probably rename these methods as it's a bit confusing now 

`Accepted` is now filtered twice in two functions, we could have high-level exported `FilterServers` which calls a chain of match functions for each server. 

```go

for _, server := range servers {
  matches := true
  
  for _, filter := range []FilterFunction{isAccepted,selectorMatches,qualifiersMatch, ... } {
     if !filter(server) {
         matches = false
         break
     }
     
     if matches {
         result = append(result, server)
      }
}
```

Reply:
Implemented something like this. Thanks.